### ISSUE TITLE
Support Java 8 time API's

### ISSUE THREAD

**Issue by semenoh** [realm/realm-java#5104]
#### Goal

> I'd like to be able to restore `java.util.Date` instance from realm db with time offset value it was stored.

#### Expected Results

> `java.util.Date` object has the same time offset as it was at the moment of writing it to realm db.

#### Actual Results

>  `java.util.Date` object has zero time offset (+00:00) even if it was non zero (for instance +03:00) at the moment of writing it to realm db.

#### Steps & Code to Reproduce

> parse "2017-08-14T12:33:00.000+03:00" string into `java.util.Date` property of RealmObject using `java.text.SimpleDateFormat.("yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ", Locale.US)

> perform `copyToRealm` for that RealmObject

> read this object from realm

> write to console this date using `java.text.SimpleDateFormat.("yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ", Locale.US).format(data.time)

> you will get date time with zero time offset
#### Code Sample

This code will not run as it is but I've got an idea

```kotlin
import java.util.*
import java.text.SimpleDateFormat

open class DateWrapper: RealmObject() {
    var date: Date? = null
}

fun main(args: Array<String>) {
    val pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
    val formatter = SimpleDateFormat(pattern, Locale.US)

    val zonedDateTimeString = "2017-08-14T12:33:00.000+0300"

    val date: Date = formatter.parse(zonedDateTimeString)
    println(formatter.format(date))
    //2017-08-14T12:33:00.000+0300
    
    val dw = DateWrapper()
    dw.date = date
    
    Realm.getDefaultInstance().use {
        it.executeTransaction {
            it.copyToRealm(dw)
        }
    }

    Realm.getDefaultInstance().use {
        println(formatter.format(it.where(DateWrapper::class.java).findFirst().date)
        // 2017-08-14T12:33:00.000+0000
    }
}

```

#### Version of Realm and tooling
Realm version(s): 3.1.4

Android Studio version: 3.0 alpha9

**Comment by cmelchior** [comment]
This is by design. Timestamps are stored in Realm in GMT (calling `date.getTime()`). If you want to store the time zone for presentation purposes, it needs to be stored in a separate field and you can then convert the Date in a custom getter, but you will need to use `Calendar` as `Date` does not enable you to set time zones.

We have talked about adding support for more timezone aware datatypes, mostly Java 8 time API's. But that is still a bit further out.

**Comment by semenoh** [comment]
`Date` does not contain any data about timeZone but it contains data about timeOffset (Although there is no way to set it except parse it from string, at least I did not found)

The reason of storing timeZone/timeOffset is not only presentation purpose but sorting and filtering to. We are getting wrong results of queries filtered by time due to appearance of records that has different date because they are in different time zone.

Unfortunately java 8 is not for the rescue as it is not available on android. So we are forced to use `three-ten-bp` or `joda-time`.

> We have talked about adding support for more timezone aware datatypes, mostly Java 8 time API's. But that is still a bit further out.

Is there any issue I can track to keep myself up to date ?

**Comment by cmelchior** [comment]
> The reason of storing timeZone/timeOffset is not only presentation purpose but sorting and filtering too. We are getting wrong results of queries filtered by time due to the appearance of records that have different date because they are in different time zone.

Yes, I could imagine. But `date` and `time` are two different concepts. Which is also the reason that Java 8 has `LocalDate` and `LocalDateTime`. `Date` is a really awful class to describe `LocalDate` semantics.

If you want to query for specific dates, I would recommend storing the date (`dd-mm-yyyy`) representation in a specific field instead as trying to calculate the correct `Date` values as doing that is neigh impossible due to the rules for timezones/daylight savings.

I don't think we have a public issue for tracking Java 8 time support, so I guess we can promote this one.

**Comment by jtoselli** [comment]
Are you planning to release something about this in the near future?

**Comment by omidraha** [comment]
Hi,
I have the same issue, I want to save only `time`, and not the `date` part of it, something like this:
```
'14:30:59'
'14:30'
```
And I would like to be able to sort query with this field.
Also there is a duplicated closed issue [here](https://github.com/realm/realm-java/issues/6476),

**Comment by asadwaheed1** [comment]
So is java.time.LocalDate support added already ?
